casper/protocols/sharding/sharding_view.py

```diff
@@ -0,0 +1,152 @@
+"""The sharding view module extends a view for sharded blockchain data structure"""
+import random as r
+
+from casper.safety_oracles.clique_oracle import CliqueOracle
+from casper.abstract_view import AbstractView
+import casper.protocols.sharding.forkchoice as forkchoice
+
+
+class ShardingView(AbstractView):
+    """A view class that keeps track of children, latest_messages_on_shard and starting_blocks"""
+    def __init__(self, messages=None, shard_genesis_block=None):
+        self.children = dict()
+
+        self.shard_genesis_blocks = dict()  # shard_id -> genesis for shard
+        self.starting_blocks = dict()  # shard_id -> starting block for forkchoice
+        self.latest_messages_on_shard = dict()  # shard_id -> validator -> message
+
+        self.select_shards = self.select_random_shards
+
+        if shard_genesis_block:
+            for shard_id in shard_genesis_block.estimate['shard_ids']:
+                self.shard_genesis_blocks[shard_id] = shard_genesis_block
+                self.starting_blocks[shard_id] = shard_genesis_block
+
+        super().__init__(messages)
+
+    def estimate(self):
+        """Returns the current forkchoice in this view"""
+        shards_forkchoice = dict()
+
+        for shard_id in sorted(self.starting_blocks):
+            shard_tip = forkchoice.get_shard_fork_choice(
+                self.starting_blocks[shard_id],
+                self.children,
+                self.latest_messages_on_shard[shard_id],
+                shard_id
+            )
+            shards_forkchoice[shard_id] = shard_tip
+
+            left_child_shard = shard_id + '0'
+            right_child_shard = shard_id + '1'
+            if left_child_shard in self.starting_blocks:
+                self.set_child_starting_block(shard_tip, shard_id, left_child_shard)
+            if right_child_shard in self.starting_blocks:
+                self.set_child_starting_block(shard_tip, shard_id, right_child_shard)
+
+        self.check_forkchoice_atomicity(shards_forkchoice)
+
+        shards_to_build_on = self.select_shards(shards_forkchoice)
+        return {'prev_blocks': {shards_forkchoice[shard_id] for shard_id in shards_to_build_on},
```

Comment:
Why are we only returning the prev_blocks for the `shards_to_build_on`? Might a user want to see the entire `shards_forkchoice` as that is the full current estimate?

Reply:
As this is the previous block links, we only point to blocks that we are explicitly building on top of w/ this new block (at most, two shards as of now). 

We can add other select_shards rules in the future to modify this slightly, but in all cases I think the prev_blocks themselves will just be those of the shards selected.